src/bms/model/control_flow.rs

```diff
@@ -0,0 +1,724 @@
+
+    fn try_from(value: Token<'a>) -> Result<Self, Self::Error> {
+        Self::try_from_token(value)
+    }
+}
+
+/// Alias preferred by external APIs/tests.
+pub type NonControlFlowToken<'a> = NonControlToken<'a>;
+
+/// A unit of branch content that can represent nested control flow or plain non-control tokens.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum TokenUnit<'a> {
+    /// Nested random block content.
+    Random(Random<'a>),
+    /// Nested switch block content.
+    Switch(Switch<'a>),
+    /// Plain non-control-flow tokens for branch content.
+    Tokens(Vec<NonControlToken<'a>>),
+}
+
+impl<'a> TokenUnit<'a> {
+    /// Create a `Tokens` unit from an iterator of raw tokens, filtering out control-flow ones.
+    #[must_use]
+    pub fn from_tokens<T>(tokens: T) -> Self
+    where
+        T: IntoIterator<Item = Token<'a>>,
+    {
+        let v = tokens
+            .into_iter()
+            .map(NonControlToken::try_from_token)
+            .flat_map(Result::ok)
+            .collect();
+        Self::Tokens(v)
+    }
+
+    /// Convert this unit into lex tokens.
+    #[must_use]
+    pub fn into_tokens(self) -> Vec<Token<'a>> {
+        match self {
+            TokenUnit::Random(r) => r.into_tokens(),
+            TokenUnit::Switch(s) => s.into_tokens(),
+            TokenUnit::Tokens(v) => v.into_iter().map(Token::from).collect(),
+        }
+    }
+}
+
+impl<'a> From<Random<'a>> for TokenUnit<'a> {
+    fn from(value: Random<'a>) -> Self {
+        TokenUnit::Random(value)
+    }
+}
+
+impl<'a> From<Switch<'a>> for TokenUnit<'a> {
+    fn from(value: Switch<'a>) -> Self {
+        TokenUnit::Switch(value)
+    }
+}
+
+impl<'a> From<Vec<NonControlToken<'a>>> for TokenUnit<'a> {
+    fn from(value: Vec<NonControlToken<'a>>) -> Self {
+        TokenUnit::Tokens(value)
+    }
+}
+
+/// Indicates whether the random block generates a value (`#RANDOM`) or uses a set value (`#SETRANDOM`).
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ControlFlowValue {
+    /// Use a fixed value (emits `#SETRANDOM <value>`).
+    Set(BigUint),
+    /// Generate a random value with the given maximum (emits `#RANDOM <max>`).
+    GenMax(BigUint),
+}
+
+/// One branch node in an if-chain inside a random block.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum IfChainEntry<'a> {
+    /// An `#ELSEIF <cond>` branch with its units and a pointer to the next entry.
+    ElseIf {
+        /// Condition value for `#ELSEIF <cond>`.
+        cond: BigUint,
+        /// Content units for this branch.
+        units: Vec<TokenUnit<'a>>,
+        /// Pointer to the next chain entry.
+        next: Box<IfChainEntry<'a>>,
+    },
+    /// An `#ELSE` branch with its units and a pointer to the next entry.
+    Else {
+        /// Content units for this branch.
+        units: Vec<TokenUnit<'a>>,
+        /// Pointer to the next chain entry.
+        next: Box<IfChainEntry<'a>>,
+    },
+    /// Terminator for an if-chain.
+    EndIf,
+}
```

Comment:
I don't want it to be public, because there should be a invariant that `IfChainEntry::Else`'s `next` field must not be a `IfChainEntry::ElseIf`.
```suggestion
/// One branch node in an if-chain inside a random block.
#[derive(Debug, Clone, PartialEq, Eq)]
enum IfChainEntry<'a> {
    /// An `#ELSEIF <cond>` branch with its units and a pointer to the next entry.
    ElseIf {
        /// Condition value for `#ELSEIF <cond>`.
        cond: BigUint,
        /// Content units for this branch.
        units: Vec<TokenUnit<'a>>,
        /// Pointer to the next chain entry.
        next: Box<IfChainEntry<'a>>,
    },
    /// An `#ELSE` branch with its units and a pointer to the next entry.
    Else {
        /// Content units for this branch.
        units: Vec<TokenUnit<'a>>,
        /// Pointer to the next chain entry.
        next: Box<IfChainEntry<'a>>,
    },
    /// Terminator for an if-chain.
    EndIf,
}
```
Also we need to check and maintain it at `append_to_tail`.